In [1]:
# Preliminaries needed for Python and to run R within a Python notebook
import rpy2.rinterface
%load_ext rpy2.ipython 

## $\chi^2$-test of Independence to Compare Multiple Proportions

Sometimes it is of interest to compare multiple experimental conditions simultaneously. When the metric being compared is a proportion, and a hypotheses of equality is being tested, we may use the $\chi^2$-test of independence to do so.

Symbolically speaking, this is the hypothesis we are testing: $$H_0: \pi_1 = \pi_2 = \cdots = \pi_m \text{ vs. } H_A: \pi_j \neq \pi_k \text{ for some } j \neq k$$
We do so by extending the ordinary two-condition $\chi^2$-test by constructing a $2 \times m$ contingency table rather than a $2 \times 2$ one. As before we compare what is _expected_ with what was _observed_. Below this process is illustrated in both `R` and `Python`. We do this is in the context of the _Nike SB_ example.

Recall that in this example there were 5 ads and there were respectively 160, 96, 141, 293 and 197 conversions for each of them. The sample size in each condition was respectively 5014, 4971, 5030, 5007 and 4980. The corresponding $2 \times 5$ contingency table is shown below.

|                |  Ad1 |  Ad2 |  Ad3 |  Ad4 |  Ad5 |       |
|---------------:|:----:|:----:|:----:|:----:|:----:|:-----:|
|     Conversion |  160 |  95  |  141 |  293 |  197 |  886  |
| Non-Conversion | 4854 | 4876 | 4889 | 4714 | 4783 | 24116 |
|                | 5014 | 4971 | 5030 | 5007 | 4980 | 25002 | 

### First in Python
Like with the $2 \times 2$ case we use the `chi2_contingency()` function in the `scipy` package. But this time we input the body (i.e., numbers of conversions and non-conversions) associated with our $2 \times 5$ table shown above.

In [9]:
## Import necessary packages
import scipy.stats as sp

## Perform the test
data_nike = [[160, 95, 141, 293, 197], [4854, 4876, 4889, 4714, 4783]]
t, p, df, exp = sp.chi2_contingency(data_nike)
print("Test statistic =", t, ", df =", df, ", p-value =", p)

Test statistic = 129.168563084 , df = 4 , p-value = 5.86411763914e-27


This is an extremely large test statistic value which, in the context of a $\chi^2$-distribution with 4 degrees of freedom yields a p-value of $5.86 \times 10^{-27}$. As such, we reject the null hypothesis and conclude that the conversion rates for each of the 5 ads is not the same.

### Next in R

As in the $2 \times 2$ case we use the `prop.test()` function. This time we input the number of conversions in each condition and the sample size in each condition (i.e., the first and third rows of numbers from the table above).

In [11]:
%%R
## Perform the test
prop.test(x = c(160, 95, 141, 293, 197), n = c(5014, 4971, 5030, 5007, 4980))


	5-sample test for equality of proportions without continuity
	correction

data:  c(160, 95, 141, 293, 197) out of c(5014, 4971, 5030, 5007, 4980)
X-squared = 129.17, df = 4, p-value < 2.2e-16
alternative hypothesis: two.sided
sample estimates:
    prop 1     prop 2     prop 3     prop 4     prop 5 
0.03191065 0.01911084 0.02803181 0.05851807 0.03955823 



We see that these results agree exactly with what we found in `Python`.